### ISSUE TITLE
move non-benchmarks (correctness tests) to different server

### ISSUE THREAD

**Issue by zookozcash** [zcash/zcash#1075]

**Comment by zookozcash** [comment]
Next step: Taylor decide whether to make the new server be EC2 or what.

Then, @arcalinea change the buildbot config (make new buildslave to represent the new server, make separate builders to represent the job of benchmarks vs the job of correctness tests)

**Comment by defuse** [comment]
EC2 m4.2xlarge (32GiB RAM, 8 vCPUs, 26 ECUs) - $103/month with a 1-year commitment, half paid up front.

**Comment by defuse** [comment]
An "ECU" is roughly a 1.0-1.2 GHz 2007 Xeon core, so the m4.2xlarge is roughly 8 cores at 26/8 = 3.25GHz.

**Comment by defuse** [comment]
SoYouStart E5-SAT-1 > EC2 m4.2xlarge (better performance and price).

**Comment by ebfull** [comment]
This ticket somewhat overlaps with #1068.

I think we need a complete overhaul of the CI stuff, and I'm not sure what ticket to post this in, but let's just do it here.

I think we need a lightweight, cheap **ci.z.cash** EC2 instance with buildbot master, homu, and coverage logs.

After that, I think it would be worthwhile to ditch our old test server by finding an EC2 instance with equivalent performance to avoid corrupting our speed test historical accuracy. The more things on EC2 the better, because it's less for us to maintain and renew all the time. But if we can't do that, I suppose we should stay with the old server.

I also believe we can use **latent buildslaves** with buildbot (see http://docs.buildbot.net/current/manual/cfg-buildslaves.html#latent-buildslaves) to spin up EC2 instances when we need them testing, which will save us some money *and* make it easier to maintain our CI infrastructure.

**Comment by defuse** [comment]
+1, except I don't think we should try to match the new server's performance to the old one, we can just re-run the performance measurements for all of our releases, which wouldn't take *too* long.

**Comment by defuse** [comment]
The server running the performance tests doesn't need a lot of cores, it just needs to be fast sequentially and have enough memory to run one instance of the performance tests (so we could use the current server for that, for now). The acceptance test and coverage stuff can be run in parallel (and parallelism exists within them e.g. `make -j8`) so we want a high-CPU server for that. 

If the acceptance test and coverage tests each use less than 8GB RAM total, 16GiB RAM is sufficient for parallelizing what we're doing now. If we want to allow for 2 more parallel builds in the future, then 32GiB RAM is sufficient.

Based on those requirements I think EC2 m4.2xlarge is better than all the in-stock SoYouStart offerings. If we're willing to spend more ($200/month) then c4.4xlarge.

**Comment by ebfull** [comment]
Keep in mind that our RPC tests need a ton of memory.

**Comment by defuse** [comment]
UGH, EC2 prices are impossibly confusing. The monthly price I was looking at is *after* the upfront thing is paid. I should have been looking at the effective hourly.

**Comment by defuse** [comment]
Now that I understand how to correctly compute monthly prices for reserved instances, EC2 looks *waaaay* too expensive for the amount of CPU/memory our builds will need.

**Comment by ebfull** [comment]
Even adjusting for the fact that with latent buildslaves we can pay only for hours we're using the build servers?

**Comment by ebfull** [comment]
Further, latent buildslaves can be configured to use spot instances on EC2 so we can get discounts during low resource utilization on AWS.

**Comment by defuse** [comment]
No, that would probably be cheaper (<$1 per acceptance test, <$5 per performance measurement).

**Comment by defuse** [comment]
I think @arcalinea should look at latent build slaves and we should see what she thinks of setting that up. There's an extra cost in that it's harder to set up and it'll make our CI system more complicated and failure-prone, but ignoring those downsides it's probably the best option.

**Comment by ebfull** [comment]
It does also have the advantages of making our tests more deterministic. Instead of manually setting up build slaves, they're set up automatically.

**Comment by str4d** [comment]
WRT migrating performance tests to a new builder, as long as we have a series of commits with duplicated benchmarks so we can compare, that's fine with me. Then the speed instance can keep the old benchmark data for historical comparison but default to the new one. The suggestion of manually running benchmarks for all previous releases is a good one.

**Comment by defuse** [comment]
Someone from the engineering meeting (best guess @str4d) had an idea: run the performance/coverage by having it track changes to our `.latest` branch (instead of it being triggered by the same homu thing). That would simplify the buildbot config a bit.

**Comment by str4d** [comment]
Wasn't me, I wasn't there ;P

The reason they are currently sequential was because we needed to avoid resource contention on our single build server. Once we have separate servers for testing and performance benchmarks, that is no longer a concern, so yes I agree that is the best way forward. I would have branch auto run zcash like now, and branch latest trigger performance followed by coverage (so those are kept in sequential order to avoid resource contention).

This will also improve the turnaround time for merging PRs, because currently the next PR won't be started until all performance tests are done. With this split, PRs will be treated and merged solely on zcash, and the performance and coverage runs will be queued up as long as it's necessary.

**Comment by ebfull** [comment]
It was my idea. :8ball: 

A separate issue, I am a bit concerned because there seems to be zombie zcash daemons running which breaks the performance tests every once in a while. I don't understand how they're not being killed in between buildbot runs.

**Comment by defuse** [comment]
Can we close this ticket now? @ebfull

**Comment by ebfull** [comment]
I think so. Let's double check with zooko though.

**Comment by defuse** [comment]
Hey @zookozcash: We think this ticket can be closed, do you agree? My understanding of what we're doing now is that the buildbot launches a fast EC2 instance, runs the tests, then shuts it down. So we get a fast server but only pay the EC2 rate for the time the tests actually take to run (and we can easily upgrade it to a faster instance in the future if we need to).

**Comment by zookozcash** [comment]
I agree! But should we also open new tickets for "have perf/cov track .latest instead of homu" (https://github.com/zcash/zcash/issues/1075#issuecomment-232205541) and "prevent zombie zcash daemons on buildslave" (https://github.com/zcash/zcash/issues/1075#issuecomment-232236074) ?

**Comment by defuse** [comment]
https://github.com/zcash/zcash/issues/1075#issuecomment-232205541 looks done and for https://github.com/zcash/zcash/issues/1075#issuecomment-232236074 I opened #1265

**Comment by ageis** [comment]
Just read up on this (closed) ticket so I have a better idea of your CI objectives and requirements. So glad we now have good automation and DevOps practices coming to the fore ;)
